File: js/src/composites/cannabisValueCalculator.ts
import { DoseCalculator } from "../dosing/doseCalculator";
import { PriceCalculator } from "../pricing/priceCalculator";

export class CannabisValueCalculator {
  static totalMgFromWeightAndPotency(
    weightGrams: number,
    potencyPercent: number
  ): number {
    return DoseCalculator.totalMgFromPercent(weightGrams, potencyPercent);
  }

  static pricePerMgFromWeightAndPotency(
    price: number,
    weightGrams: number,
    potencyPercent: number
  ): number {
    const totalMg = DoseCalculator.totalMgFromPercent(
      weightGrams,
      potencyPercent
    );

    return PriceCalculator.pricePerMg(price, totalMg);
  }

  static pricePerGramAfterDiscount(
    price: number,
    grams: number,
    percentOff: number
  ): number {
    const discounted = PriceCalculator.discountedPrice(price, percentOff);
    return PriceCalculator.pricePerGram(discounted, grams);
  }

  static pricePerMgAfterDiscount(
    price: number,
    weightGrams: number,
    potencyPercent: number,
    percentOff: number
  ): number {
    const discounted = PriceCalculator.discountedPrice(price, percentOff);
    return this.pricePerMgFromWeightAndPotency(
      discounted,
      weightGrams,
      potencyPercent
    );
  }
}

File: src/CannabisMath.Core/Composites/CannabisValueCalculator.cs
using CannabisMath.Core.Dosing;
using CannabisMath.Core.Pricing;

namespace CannabisMath.Core.Composites;

public static class CannabisValueCalculator
{
    public static decimal TotalMgFromWeightAndPotency(decimal weightGrams, decimal potencyPercent)
    {
        return DoseCalculator.TotalMgFromPercent(weightGrams, potencyPercent);
    }

    public static decimal PricePerMgFromWeightAndPotency(
        decimal price,
        decimal weightGrams,
        decimal potencyPercent)
    {
        var totalMg = DoseCalculator.TotalMgFromPercent(weightGrams, potencyPercent);

        return PriceCalculator.PricePerMg(price, totalMg);
    }

    public static decimal PricePerGramAfterDiscount(
        decimal price,
        decimal grams,
        decimal percentOff)
    {
        var discountedPrice = PriceCalculator.DiscountedPrice(price, percentOff);

        return PriceCalculator.PricePerGram(discountedPrice, grams);
    }

    public static decimal PricePerMgAfterDiscount(
        decimal price,
        decimal weightGrams,
        decimal potencyPercent,
        decimal percentOff)
    {
        var discountedPrice = PriceCalculator.DiscountedPrice(price, percentOff);
        var totalMg = DoseCalculator.TotalMgFromPercent(weightGrams, potencyPercent);

        return PriceCalculator.PricePerMg(discountedPrice, totalMg);
    }
}